Return early if required UTXOs already big enough

tcharding opened the pull request:
### Description

If the required UTXO set is already bigger (including fees) than the
amount required for the transaction we can return early, no need to go
through the BNB algorithm or random selection.

### Notes to the reviewers

This is an optimisation, probably wants careful review since I only just started reading this code :)

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [ ] I've added tests for the new feature
* [ ] I've added docs for the new feature
* [ ] I've updated `CHANGELOG.md`

#### Bugfixes:

* [ ] This pull request breaks the existing API
* [ ] I've added tests to reproduce the issue which are now passing
* [ ] I'm linking the issue being fixed by this PR

tcharding committed:
Return early if required UTXOs already big enough

If the required UTXO set is already bigger (including fees) than the
amount required for the transaction we can return early, no need to go
through the BNB algorithm or random selection.

afilini commented:
I can't think of anything that would break because of this, but just to be extra careful: could you also add a test? If anything just to make sure our test coverage also goes through that branch.

tcharding commented:
> I can't think of anything that would break because of this, but just to be extra careful: could you also add a test? If anything just to make sure our test coverage also goes through that branch.

Sure thing, can do. My bad, should have done this already.

tcharding committed:
Add unit test required not enough

Add a unit test that passes a required utxo to the coin selection
algorithm that is less than the required spend. This tests that we get
that utxo included as well as tests that the rest of the coin selection
algorithm code also executes (i.e., that we do not short circuit
incorrectly).

tcharding commented:
Turns out this code path is already tested by the existing test `test_bnb_coin_selection_required_are_enough`, I've added an additional test `test_bnb_coin_selection_required_not_enough` to test for falsely entering the new code path.
